common/plugins/eu.esdihumboldt.hale.common.instance.index/src/eu/esdihumboldt/hale/common/instance/index/InstanceIndexServiceImpl.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) 2017 wetransform GmbH
+ * 
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution. If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Contributors:
+ *     wetransform GmbH <http://www.wetransform.to>
+ */
+
+package eu.esdihumboldt.hale.common.instance.index;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.xml.namespace.QName;
+
+import eu.esdihumboldt.hale.common.align.extension.transformation.TransformationFunctionUtil;
+import eu.esdihumboldt.hale.common.align.extension.transformation.TypeTransformationFactory;
+import eu.esdihumboldt.hale.common.align.model.AlignmentUtil;
+import eu.esdihumboldt.hale.common.align.model.Cell;
+import eu.esdihumboldt.hale.common.align.model.impl.PropertyEntityDefinition;
+import eu.esdihumboldt.hale.common.align.transformation.function.TypeTransformation;
+import eu.esdihumboldt.hale.common.core.service.ServiceProvider;
+import eu.esdihumboldt.hale.common.instance.model.Identifiable;
+import eu.esdihumboldt.hale.common.instance.model.IdentifiableInstanceReference;
+import eu.esdihumboldt.hale.common.instance.model.Instance;
+import eu.esdihumboldt.hale.common.instance.model.InstanceCollection;
+import eu.esdihumboldt.hale.common.instance.model.InstanceReference;
+import eu.esdihumboldt.hale.common.instance.model.ResolvableInstanceReference;
+import eu.esdihumboldt.hale.common.instance.model.impl.DefaultInstanceCollection;
+
+/**
+ * Maintains instance indexes
+ * 
+ * @author Florian Esser
+ */
+public class InstanceIndexServiceImpl implements InstanceIndexService {
+
+	private final Map<QName, MultimapInstanceIndex> indexes = new HashMap<>();
```

Comment:
Is there a specific reason why the concrete `InstanceIndex` implementation is used here and not the interface?

Reply:
No specific reason, I refactored it so that now the interface `HaleInstanceIndex` is used. Always using `InstanceIndex` directly is a bit cumbersome because of the generics. Definitely could be refactored further in the future in case we want to use another index implementation.